<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">  
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>Running DegNorm - DegNorm</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/cinder.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            <a class="navbar-brand" href="../..">DegNorm</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">How-Tos <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../installation/">Installation</a>
</li>

                        
                            
<li class="active">
    <a href="./">Running DegNorm</a>
</li>

                        
                            
<li >
    <a href="../speedups/">Speed enhancements</a>
</li>

                        
                            
<li >
    <a href="../posthoc_analysis/">Posthoc analysis</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">About <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../about/license/">License</a>
</li>

                        
                            
<li >
    <a href="../../about/release/">Release notes</a>
</li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fa fa-search"></i> Search
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../installation/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../speedups/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#running-degnorm">Running DegNorm</a></li>
            <li class="second-level"><a href="#inputs">Inputs</a></li>
                
                <li class="third-level"><a href="#1-align-and-sort-reads-data">1. Align and sort reads data</a></li>
                <li class="third-level"><a href="#2-paired-or-single-end-aligned-and-sorted-reads-data">2. (Paired or single end) aligned and sorted reads data</a></li>
                <li class="third-level"><a href="#3-genome-annotation-file">3. Genome annotation file</a></li>
            <li class="second-level"><a href="#using-a-warm-start-directory">Using a warm start directory</a></li>
                
            <li class="second-level"><a href="#runtime-and-algorithmic-parameters">Runtime and algorithmic parameters</a></li>
                
            <li class="second-level"><a href="#example-usage">Example usage</a></li>
                
                <li class="third-level"><a href="#required-sorting-indexing-alignment-files">Required: sorting, indexing alignment files</a></li>
                <li class="third-level"><a href="#launching-a-full-degnorm-run-the-verbose-way">Launching a full degnorm run "the verbose way"</a></li>
                <li class="third-level"><a href="#launching-a-full-degnorm-run-the-easy-way">Launching a full degnorm run "the easy way"</a></li>
                <li class="third-level"><a href="#using-a-warm-start-directory_1">Using a warm start directory</a></li>
                <li class="third-level"><a href="#launching-distributed-runs-with-mpi">Launching distributed runs with MPI</a></li>
            <li class="second-level"><a href="#output">Output</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="running-degnorm">Running DegNorm</h1>
<p>DegNorm is a command line interface (CLI) tool. You can access it through the <code>degnorm</code> command and the MPI-enabled
distributed version is <code>degnorm_mpi</code>.</p>
<hr />
<h2 id="inputs">Inputs</h2>
<p>You only need two types of files to supply <code>degnorm</code> - .bam files and a .gtf file
if you have <code>samtools</code> in your <code>$PATH</code>. If you do not have <code>samtools</code>, you will also need <a href="[bam index files](https://www.biostars.org/p/15847/)">bam index files</a> to accompany your .bam files.</p>
<h4 id="1-align-and-sort-reads-data">1. Align and sort reads data</h4>
<p>Before running DegNorm, you will need to sort (and optionally, index) them with <code>samtools sort</code>. This command just re-orders reads
by chromosome and starting index, so that they can be more useful to DegNorm later on.</p>
<p>To <a href="https://davetang.org/wiki/tiki-index.php?page=SAMTools#Sorting_a_BAM_file">sort</a> an alignment file, suppose it's called <code>S1.bam</code>, you can sort it with the following command:</p>
<pre><code>samtools sort S1.bam -o S1_sorted.bam
</code></pre>
<p>DegNorm requires sorted .bam files because it really needs <a href="https://www.biostars.org/p/15847/">bam index files</a>. If you don't create them
first, DegNorm will run <code>samtools index</code> to create an index file for each input alignment file:</p>
<pre><code>samtools index S1_sorted.bam S1_sorted.bai
</code></pre>
<h4 id="2-paired-or-single-end-aligned-and-sorted-reads-data">2. (Paired or single end) aligned and sorted reads data</h4>
<p>At least two alignment files (.bam) must be specified, as inter-sample degradation normalization cannot happen on a standalone 
RNA-Seq sample. Use <code>p</code> to refer to the total number of samples.</p>
<p>If .bai files are not submitted, <code>degnorm</code> will look for .bai files named after the .bam files only with the ".bai" extension.
If no such file is found, <code>degnorm</code> will attempt to build one with the <code>samtools index</code> command. This will only work if the .bam files are sorted.
Instead of specifying individual .bam and .bai files, you can just specify <code>--bam-dir</code>, a path to a directory holding the relevant .bam and .bai files.
  With <code>--bam-dir</code>, it is assumed that the .bai files are named according to the .bam files, only with the ".bai" extension.</p>
<p><strong>You can supply paired reads files or single end reads files</strong>.</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--bam-files</code></td>
<td>If neither <code>--warm-start-dir</code> nor <code>--bam-dir</code> are specified</td>
<td>Set of individual .bam files</td>
</tr>
<tr>
<td><code>--bai-files</code></td>
<td>If <code>samtools</code> is not in the <code>$PATH</code> and neither <code>--warm-start-dir</code> nor <code>--bam-dir</code> are specified</td>
<td>Set of individual .bai files. If specified, they must be in the order corresponding to <code>--bam-files</code>.</td>
</tr>
<tr>
<td><code>--bam-dir</code></td>
<td>If neither <code>--warm-start-dir</code> nor <code>--bam-files</code> are specified</td>
<td>Directory containing .bam and .bai files for a pipeline run. It is assumed the .bai files have the same name as the .bam files, just with a different extension.</td>
</tr>
<tr>
<td><code>-u</code>, <code>--unique-alignments</code></td>
<td>optional flag</td>
<td>If specified, tells <code>degnorm</code> to remove reads aligned to more than one location on the genome. Suggested for use with single end reads data.</td>
</tr>
</tbody>
</table>
<h4 id="3-genome-annotation-file">3. Genome annotation file</h4>
<p>DegNorm needs a .gtf file in order to construct the total transcript and compute all per-gene coverage score curves. It is assumed your .gtf file abides by the standard <a href="https://useast.ensembl.org/info/website/upload/gff.html">conventions</a>.</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-g</code>, <code>--genome-annotation</code></td>
<td>Yes</td>
<td>.gtf file for relevant genome.</td>
</tr>
</tbody>
</table>
<h2 id="using-a-warm-start-directory">Using a warm start directory</h2>
<p>Loading multiple .bam files, parsing a .gtf file, and computing per-gene cross-sample coverage matrices and read counts, can take some time, but this is a one-time
 preprocessing cost given a specific set of RNA-Seq samples. If you run the <code>degnorm</code> pipeline once, you can leverage
the stored coverage, read counts, and parsed transcript data from an existing DegNorm output directory to start a a new DegNorm run where all of the
preprocessing is completed ahead of time. When using <code>--warm-start-dir</code> you do <em>not</em> need to supply a .gtf file.</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--warm-start-dir</code></td>
<td>No</td>
<td>A directory holding the contents of a successful previous <code>degnorm</code> pipeline run. When specified, there is no need to supply a .gtf file.</td>
</tr>
</tbody>
</table>
<h2 id="runtime-and-algorithmic-parameters">Runtime and algorithmic parameters</h2>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-o</code>, <code>--output-dir</code></td>
<td>No</td>
<td>Defaults to the current working directory. Use to specify location where pipeline output directory will be written.</td>
</tr>
<tr>
<td><code>--plot-genes</code></td>
<td>No</td>
<td>Names of genes for which to render coverage plots. Sequence of explictly stated gene names or a .txt file containing one gene name per line.</td>
</tr>
<tr>
<td><code>-d</code>, <code>--downsample-rate</code></td>
<td>No</td>
<td>Integer downsampling rate. Systematic samples of a coverage matrix are used to speed up NMF iterations.</td>
</tr>
<tr>
<td><code>--nmf-iter</code></td>
<td>No</td>
<td>Number of iterations per NMF-OA approximation. The higher the more accurate the approximation, but the more costly in terms of time.</td>
</tr>
<tr>
<td><code>--iter</code></td>
<td>No</td>
<td>Number of whole DegNorm iterations. Default is 5.</td>
</tr>
<tr>
<td><code>--minimax-coverage</code></td>
<td>No</td>
<td>Minimum cross-sample maximum coverage for a gene before it is included in the DegNorm pipeline. Can be used to exclude relatively low-coverage genes.</td>
</tr>
<tr>
<td><code>-s</code>, <code>--skip-baseline-selection</code></td>
<td>No</td>
<td>Flag to skip baseline selection, will greatly speed up DegNorm iterations.</td>
</tr>
<tr>
<td><code>-u</code>, <code>--unique-alignments</code></td>
<td>No</td>
<td>Flag, only keep uniquely mapped reads (reads with <code>NH</code> (number of hits) == 1)</td>
</tr>
<tr>
<td><code>-p</code>, <code>--proc-per-node</code></td>
<td>No</td>
<td>Integer number of processes to spawn per compute node. The more the better.</td>
</tr>
</tbody>
</table>
<h2 id="example-usage">Example usage</h2>
<h3 id="required-sorting-indexing-alignment-files">Required: sorting, indexing alignment files</h3>
<p>Remember - prior to running <code>degnorm</code>, you must <code>sort</code> your aligned reads and create index files (.bai files).
Suppose we have set of .bam files in our current working directory that we would like to send through DegNorm, the following
commands will sort the .bam files and create .bai files, all with <code>samtools</code>:</p>
<pre><code># sort and index alignment files in current working directory.
for FILE in ./*.bam
do
    echo 'sorting '$FILE
    samtools sort $FILE -o ${FILE/.bam/}'_sorted.bam'
    echo 'indexing '${FILE/.bam/}'_sorted.bam'
    samtools index ${FILE/.bam/}'_sorted.bam' ${FILE/.bam/.bai}
done
</code></pre>
<h3 id="launching-a-full-degnorm-run-the-verbose-way">Launching a full <code>degnorm</code> run "the verbose way"</h3>
<p>Next, after the alignment files have been sorted (and optionally indexed), let's run <code>degnorm</code> "the verbose way" by enumerating 
each .bam and .bai file individually. Suppose we have two sorted alignment files, <code>S1_sorted.bam</code> and <code>S2_sorted.bam</code>.</p>
<pre><code># run degnorm on two samples on 20 cores, using 100 NMF iterations per gene
degnorm --bam-files S1_sorted.bam S2_sorted.bam \
    --bai-files S1_sorted.bai S2_sorted.bai \
    -g human.gtf \
    -p 20 \
    -o degnorm_output
</code></pre>
<p>If we did not specify any <code>--bai-files</code>, then DegNorm would automatically search for files "S1_sorted.bai" and "S2_sorted.bai" because they're named after
the alignment files, "S1_sorted.bam" and "S2_sorted.bam." If those index files cannot be found, they will be created in the same directory
  containing the input .bam files, only if <code>samtools</code> is in your <code>$PATH</code>.</p>
<h3 id="launching-a-full-degnorm-run-the-easy-way">Launching a full <code>degnorm</code> run "the easy way"</h3>
<p>It's probably just easier to supply <code>degnorm</code> with an input directory containing all of our .bam files, instead of each .bam file individually.
Suppose that data directory is located at <code>degnorm_data/GBM</code>.
 Let's also use 5 DegNorm iterations, 50 NMF iterations per gene per NMF iteration, and route output to a directory besides the current working directory.</p>
<pre><code>degnorm --bam-dir degnorm_data/GBM \
    -g human.gtf \
    -p 20 \
    --nmf-iter 50 \
    -o degnorm_output
</code></pre>
<h3 id="using-a-warm-start-directory_1">Using a warm start directory</h3>
<p>After one pipeline run, we could start <code>degnorm</code> from the output directory resulting from the first run - this directory is known as the "warm start" directory. This time, let's exclude genes with a maximum coverage
(across all samples) less than 20, and run DegNorm with only 3 iterations. In this example, <code>DegNorm_102118_081045</code> was generated automatically from one of the prior DegNorm runs.</p>
<pre><code>degnorm --warm-start-dir degnorm_output/DegNorm_102118_081045 \
    -p 20 \
    -o degnorm_output \
    --minimax-coverage 20 \
    --iter 3
</code></pre>
<h3 id="launching-distributed-runs-with-mpi">Launching distributed runs with MPI</h3>
<p>Let's speed this up by distributing DegNorm's workload over multiple servers with <code>degnorm_mpi</code>!
 If the <code>mpi4py</code> python package has been installed (meaning that MPICH is available in your compute environment).
 You can still specify the number of cores to use on each node with <code>-p</code>.</p>
<pre><code># run degnorm_mpi using 4 nodes ("-n 4")
mpiexec -n 4 degnorm_mpi \
    --warm-start-dir degnorm_output/DegNorm_GBM_102018 \
    -p 20 \
    -o degnorm_output \
    --minimax-coverage 20
    --nmf-iter 100
</code></pre>
<p>The functionality of <code>degnorm_mpi</code> is the same as the single-node <code>degnorm</code>, so you can continue using warm start directories just like you could with <code>degnorm</code>.
 If you are unfamiliar with MPI, a great </p>
<h2 id="output">Output</h2>
<p>Raw and estimated coverage data are stored in separate <code>.pkl</code> files, one file per chromosome. See the <a href="../posthoc_analysis/">posthoc analysis</a>
 documentation for helper functions to access coverage data.</p>
<p>In addition to per-gene coverage matrices, <code>degnorm</code> will produce raw and degradation-adjusted read counts, a matrix of degradation index scores,
a matrix describing which genes were sent through the baseline selection procedure and when, along with various summary graphics and a pipeline summary report. 
The report will render as an .html file, but if you have <code>pandoc</code> installed
it will be converted to a .pdf.</p>
<pre><code>├── degnorm.log
├── degradation_index_scores.csv
├── ran_baseline_selection.csv  # matrix of Booleans: which genes go thru baseline selection
├── gene_exon_metadata.csv  # chromosome, gene, exon relationship data
├── read_counts.csv  # raw read counts
├── adjusted_read_counts.csv  # degradation normalized read counts
├── &lt;SAMPLE_1&gt;
│   ├── sample_SAMPLE_1_chr1.npz  # full chrom coverage for sample 1 in compressed matrix.
│   ├── sample_SAMPLE_1_chr2.npz
│   ├── &lt;more sample chromosome coverage .npz files&gt;
├── &lt;SAMPLE_2&gt;
│   ├── sample_SAMPLE_2_chr1.npz
│   ├── sample_SAMPLE_2_chr2.npz
│   ├── &lt;more per-sample chromosome coverage .npz files&gt;
├── &lt;more sample directories with chromosome coverage&gt;
├── chr1
│   ├── GAPDH_coverage.png
│   ├── &lt;more coverage plots&gt;
│   ├── coverage_matrices_chr1.pkl  # (gene, raw coverage matrix) pairs in serialized python object.
│   └── estimated_coverage_matrices_chr1.pkl # (gene, estimated coverage matrix) pairs in serialized python object.
├── chr2
│   ├── coverage_matrices_chr2.pkl
│   └── estimated_coverage_matrices_chr2.pkl
├── chr3
│   ├── coverage_matrices_chr3.pkl
│   └── estimated_coverage_matrices_chr3.pkl
├── chr4
│   ├── GAPDH_coverage.png
│   ├── coverage_matrices_chr4.pkl
│   └── estimated_coverage_matrices_chr4.pkl
├── &lt;more chromosome directories&gt;
│   ├── &lt;more raw coverage matrix .pkl files&gt;
│   └── &lt;more estimated coverage matrix .pkl files&gt;
└── report
    ├── degnorm_summary.pdf # (or .html if pandoc not available)
    ├── di_boxplots.png
    ├── di_correlation.png
    └── di_heatmap.png
</code></pre></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script src="../../search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
