<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">  
    <link rel="shortcut icon" href="../../img/favicon.ico">

    <title>Running DegNorm - DegNorm</title>

    <link href="../../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../../css/base.css" rel="stylesheet">
    <link href="../../css/cinder.css" rel="stylesheet">
    <link rel="stylesheet" href="../../css/highlight.css">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            <a class="navbar-brand" href="../..">DegNorm</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../..">Home</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">How-Tos <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../installation/">Installation</a>
</li>

                        
                            
<li class="active">
    <a href="./">Running DegNorm</a>
</li>

                        
                            
<li >
    <a href="../posthoc_analysis/">Posthoc Analysis</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">About <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../about/license/">License</a>
</li>

                        
                            
<li >
    <a href="../../about/release/">Release Notes</a>
</li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fa fa-search"></i> Search
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../installation/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../posthoc_analysis/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#running-degnorm">Running DegNorm</a></li>
            <li class="second-level"><a href="#inputs">Inputs</a></li>
                
                <li class="third-level"><a href="#1-paired-or-single-end-aligned-and-sorted-reads-data">1. (Paired or single end) aligned and sorted reads data</a></li>
                <li class="third-level"><a href="#2-genome-annotation-file">2. Genome annotation file</a></li>
            <li class="second-level"><a href="#using-a-warm-start-directory">Using a warm start directory</a></li>
                
            <li class="second-level"><a href="#runtime-and-algorithmic-parameters">Runtime and algorithmic parameters</a></li>
                
            <li class="second-level"><a href="#example-usage">Example usage</a></li>
                
            <li class="second-level"><a href="#output">Output</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="running-degnorm">Running DegNorm</h1>
<p>DegNorm is a CLI tool. You can access it through the <code>degnorm</code> command.</p>
<h2 id="inputs">Inputs</h2>
<p>You only need two types of files to supply <code>degnorm</code>: .bam (and their corresponding .bai files - <a href="https://www.biostars.org/p/15847/">bam index files</a>), and a .gtf file.
If you have <code>samtools</code> in your <code>$PATH</code>, .bai files will be created for you if you do not have them.</p>
<h4 id="1-paired-or-single-end-aligned-and-sorted-reads-data">1. (Paired or single end) aligned and sorted reads data</h4>
<p>At least two aligned reads files must be specieid, as inter-sample degradation normalization can't happen on a standalone 
RNA-Seq expermient. Use <code>p</code> to refer to the total number of experiments.</p>
<p>It is assumed your .bam files are <strong>sorted</strong> (i.e. with <code>samtools sort</code>), contain a header, and abide by the <a href="http://samtools.sourceforge.net/SAM1.pdf">conventions</a>. If .bai files are not submitted,
<code>degnorm</code> will look for .bai files named after the .bam files only with the ".bai" extension. If no such file is found, <code>degnorm</code> will attempt to build one with the <code>samtools index</code> command. This will only work if the .bam files are sorted.
Instead of specifying individual .bam and .bai files, you can just specify <code>--bam-dir</code>, a path to a directory holding the relevant .bam and .bai files.
  With <code>--bam-dir</code>, it is assumed that the .bai files are named the same as the .bam files, they just have a different extension.</p>
<p><strong>You can supply paired reads files or single end reads files</strong>.</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--bam-files</code></td>
<td>If neither <code>--warm-start-dir</code> nor <code>--bam-dir</code> are specified</td>
<td>Set of individual .bam files</td>
</tr>
<tr>
<td><code>--bai-files</code></td>
<td>If <code>samtools</code> is not in the <code>$PATH</code> and neither <code>--warm-start-dir</code> nor <code>--bam-dir</code> are specified</td>
<td>Set of individual .bai files. If specified, they must be in the order corresponding to <code>--bam-files</code>.</td>
</tr>
<tr>
<td><code>--bam-dir</code></td>
<td>If neither <code>--warm-start-dir</code> nor <code>--bam-files</code> are specified</td>
<td>Directory containing .bam and .bai files for a pipeline run. It is assumed the .bai files are named after the .bam files.</td>
</tr>
<tr>
<td><code>-u</code>, <code>--unique-alignments</code></td>
<td>optional flag</td>
<td>If specified, tells <code>degnorm</code> to remove reads aligned to more than one location on the genome. Suggested for use with single end reads data.</td>
</tr>
</tbody>
</table>
<h4 id="2-genome-annotation-file">2. Genome annotation file</h4>
<p>DegNorm needs a .gtf file to determine the transcript for computing the per-gene coverage curves, which span a 
concatenation of the coding regions only (introns are ignored). It is assumed your .gtf file abides by the standard <a href="https://useast.ensembl.org/info/website/upload/gff.html">conventions</a>.</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-g</code>, <code>--genome-annotation</code></td>
<td>Yes</td>
<td>.gtf file for relevant genome.</td>
</tr>
</tbody>
</table>
<h2 id="using-a-warm-start-directory">Using a warm start directory</h2>
<p>Loading multiple .bam files, parsing a .gtf files, and computing per-gene cross-sample coverage matrices and read counts, can take some time, but this is a one-time
 preprocessing cost given a specific set of RNA-Seq experiments. If you run the <code>degnorm</code> pipeline once, you can leverage
the stored coverage, read counts, and parsed transcript data from an existing DegNorm output directory to start a a new DegNorm run where all of the
preprocessing is completed ahead of time. When using <code>--warm-start-dir</code> you do <em>not</em> need to supply a .gtf file.</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--warm-start-dir</code></td>
<td>No</td>
<td>A directory holding the contents of a successful previous <code>degnorm</code> pipeline run. When specified, there is no need to supply a .gtf file.</td>
</tr>
</tbody>
</table>
<h2 id="runtime-and-algorithmic-parameters">Runtime and algorithmic parameters</h2>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Required?</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-o</code>, <code>--output-dir</code></td>
<td>No</td>
<td>Defaults to the current working directory. Use to specify location where pipeline output directory will be written.</td>
</tr>
<tr>
<td><code>--plot-genes</code></td>
<td>No</td>
<td>Names of genes for which to render coverage plots. Sequence of explictly stated gene names or a .txt file containing one gene name per line.</td>
</tr>
<tr>
<td><code>-d</code>, <code>--downsample-rate</code></td>
<td>No</td>
<td>EXPERIMENTAL. Integer downsampling rate. Systematic samples of a coverage matrix are used to speed up NMF iterations.</td>
</tr>
<tr>
<td><code>--nmf-iter</code></td>
<td>No</td>
<td>Number of iterations per NMF-OA approximation. The higher the more accurate the approximation, but the more costly in terms of time.</td>
</tr>
<tr>
<td><code>--iter</code></td>
<td>No</td>
<td>Number of whole DegNorm iterations. Default is 5.</td>
</tr>
<tr>
<td><code>--minimax-coverage</code></td>
<td>No</td>
<td>Minimum cross-sample maximum coverage for a gene before it is included in the DegNorm pipeline. Can be used to exclude relatively low-coverage genes.</td>
</tr>
<tr>
<td><code>-s</code>, <code>--skip-baseline-selection</code></td>
<td>No</td>
<td>EXPERIMENTAL. Flag to skip baseline selection, will greatly speed up DegNorm iterations.</td>
</tr>
<tr>
<td><code>-c</code>, <code>--cpu</code></td>
<td>No</td>
<td>Integer number of threads. The more the better.</td>
</tr>
</tbody>
</table>
<h2 id="example-usage">Example usage</h2>
<p>Run <code>degnorm</code> with an input directory (containing multiple .bam files) with 20 threads, use 5 DegNorm iterations and 50 NMF iterations per gene per NMF iteration.
Route output to a directory besides the current working directory.</p>
<pre><code>degnorm --bam-dir degnorm_data/GBM \
    -g human.gtf \
    -c 20 \
    --nmf-iter 50 \
    -o degnorm_output
</code></pre>
<p>This is equivalent to simply enumerating the .bam files individually. Here, since .bai files are not explicitly specified, <code>degnorm</code> will look for 
the files "degnorm_data/GBM/S1.bai" and  "degnorm_data/GBM/S2.bai":</p>
<pre><code>degnorm --bam-files degnorm_data/GBM/S1.bam degnorm_data/GBM/S2.bam \
    -g human.gtf \
    -c 20 \
    --nmf-iter 50 \
    -o degnorm_output
</code></pre>
<p>After one pipeline run, we could start <code>degnorm</code> from a warm start directory (from a previous run). This time, do not include genes with a maximum coverage
(across all samples) less than 20, and only run 3 DegNorm iterations.</p>
<pre><code>degnorm --warm-start-dir degnorm_output/DegNorm_GBM_102018 \
    -c 20 \
    -o degnorm_output \
    -minimax-coverage \
    --iter 3
</code></pre>
<h2 id="output">Output</h2>
<p>Raw and estimated coverage data are stored in separate <code>.pkl</code> files, one file per chromosome. See the <a href="../posthoc_analysis/">posthoc analysis</a>
 documentation for helper functions to access coverage data.</p>
<p>In addition to per-gene coverage matrices, <code>degnorm</code> will produce raw and degradation-adjusted read counts, a matrix of degradation index scores,
a matrix describing which genes were sent through the baseline selection procedure and when, along with various summary graphics and a pipeline summary report. 
The report will render as an .html file, but if you have <code>pandoc</code> installed
it will be converted to a .pdf.</p>
<pre><code>├── degnorm.log
├── degradation_index_scores.csv
├── ran_baseline_selection.csv
├── adjusted_read_counts.csv
├── read_counts.csv
├── adjusted_read_counts.csv
├── SAMPLE_1
│   ├── sample_SAMPLE_1_chr1.npz
│   ├── sample_SAMPLE_1_chr2.npz
│   ├── &lt;more sample chromosome coverage .npz files&gt;
├── SAMPLE_2
│   ├── sample_SAMPLE_2_chr1.npz
│   ├── sample_SAMPLE_2_chr2.npz
│   ├── &lt;more per-sample chromosome coverage .npz files&gt;
├── &lt;more sample directories with chromosome coverage&gt;
├── chr1
│   ├── GAPDH_coverage.png
│   ├── &lt;more coverage plots&gt;
│   ├── coverage_matrices_chr1.pkl
│   └── estimated_coverage_matrices_chr1.pkl
├── chr2
│   ├── coverage_matrices_chr2.pkl
│   └── estimated_coverage_matrices_chr2.pkl
├── chr3
│   ├── coverage_matrices_chr3.pkl
│   └── estimated_coverage_matrices_chr3.pkl
├── chr4
│   ├── GAPDH_coverage.png
│   ├── coverage_matrices_chr4.pkl
│   └── estimated_coverage_matrices_chr4.pkl
├── &lt;more chromosome directories&gt;
│   ├── &lt;more raw coverage matrix .pkl files&gt;
│   └── &lt;more estimated coverage matrix .pkl files&gt;
└── report
    ├── degnorm_summary.pdf # (or .html if pandoc not available)
    ├── di_boxplots.png
    ├── di_correlation.png
    └── di_heatmap.png
</code></pre></div>
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../../js/jquery-1.10.2.min.js"></script>
    <script src="../../js/bootstrap-3.0.3.min.js"></script>
    <script src="../../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '../..';
    </script>
    <script data-main="../../mkdocs/js/search.js" src="../../mkdocs/js/require.js"></script>
    <script src="../../js/base.js"></script>
    <script src="../../search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
